File: lib/python-context/src/brevis/context.py
"""Read what the steps before published; publish something for the ones after.

# Why this is a module and not a class

``BrevisContext.get(...)`` as a classmethod suggests there could be two of them.
There cannot: a process is inside exactly one step of one run. The import says
that, and there is no object to pass around or to mock.

# How it reaches the engine

It does not. There is no API call, no token, no port and no network.

The engine hands a step ``BREVIS_INPUT`` when it starts it, and reads back the
file at ``BREVIS_OUTPUT`` when it ends -- under Kubernetes that path is
``/dev/termination-log``, which the engine already reads to get the exit code.
So this package is a JSON parse and a file write, which is why it is pure
standard library and why its tests need no cluster, no database and no engine.

# The ceiling

4096 bytes, and it is the platform's number rather than ours. The termination
message is TRUNCATED rather than refused, so an oversized object arrives cut
mid-string and reads downstream as "this step published nothing". Refusing
before writing is what turns that into a message.
"""

from __future__ import annotations

import atexit
import json
import logging
import os
import sys
import tempfile
from typing import Any, Mapping

__all__ = [
    "ContextError",
    "NotVisible",
    "TooLarge",
    "get",
    "of",
    "published",
    "set",
]

ENV_INPUT = "BREVIS_INPUT"
ENV_OUTPUT = "BREVIS_OUTPUT"
ENV_RUN_ID = "BREVIS_RUN_ID"

#: The prefix the engine looks for on stdout. The phases use it, the metrics use
#: it, and the context uses it when there is no file to write.
#:
#: Defined HERE and imported by :mod:`brevis.metrics` rather than written twice.
#: A protocol literal in two files is a protocol that changes in one of them.
MARKER = "@brevis:"

#: Kubernetes truncates the termination message at this size. See the module
#: docstring: inheriting the platform's ceiling is the design.
MAX_BYTES = 4096

_log = logging.getLogger("brevis")


class ContextError(Exception):
    """Anything this module refuses."""


class NotVisible(ContextError, KeyError):
    """A step was read that this one does not depend on.

    It subclasses KeyError so ``except KeyError`` around a lookup keeps working,
    and ContextError so a caller can tell it from a plain missing dict key.
    """


class TooLarge(ContextError):
    """More was published than the termination message can carry."""


# --------------------------------------------------------------------------
# Reading
# --------------------------------------------------------------------------


def _incoming() -> dict[str, dict[str, Any]]:
    raw = os.environ.get(ENV_INPUT, "").strip()
    if not raw:
        return {}
    try:
        parsed = json.loads(raw)
    except json.JSONDecodeError as exc:  # pragma: no cover - the engine writes it
        raise ContextError(
            f"{ENV_INPUT} is not valid JSON. The engine writes it, so this means a "
            f"version mismatch rather than anything you did: {exc}"
        ) from exc
    if not isinstance(parsed, dict):
        raise ContextError(f"{ENV_INPUT} has to be a JSON object, and it is {type(parsed).__name__}")
    return parsed


def get(key: str, default: Any = None) -> Any:
    """Read one value published by a step this one depends on.

        bucket = context.get("extract.bucket")

    The key is ALWAYS qualified by the step that wrote it, and a bare
    ``get("bucket")`` is refused naming the steps that published it.

    That is not ceremony. Steps are isolated by design -- ``extract`` and
    ``transform`` cannot overwrite each other, because each writes only its own
    namespace -- and a bare key throws that away the moment two of them publish
    ``bucket``. Searching and picking one is precedence by accident: it works
    until it silently does not, and somebody spends an afternoon on a value that
    came from the wrong step.

    An absent key is not an error; that is what ``default`` is for. An absent
    STEP is, because it means the pipeline does not say what the code assumes.
    """
    step, sep, field = key.partition(".")
    if not sep or not field:
        raise ContextError(_unqualified(key))

    incoming = _incoming()
    if step not in incoming:
        raise NotVisible(_not_visible(step, sorted(incoming)))
    return incoming[step].get(field, default)


def of(step: str) -> Mapping[str, Any]:
    """Everything one step published, as a read-only mapping.

    For the case where a step publishes several related values and the reader
    wants them together. It returns a copy: mutating what came in would look
    like it changed something.
    """
    incoming = _incoming()
    if step not in incoming:
        raise NotVisible(_not_visible(step, sorted(incoming)))
    return dict(incoming[step])


def _unqualified(key: str) -> str:
    owners = [s for s, values in _incoming().items() if key in values]
    if owners:
        suggestion = ", ".join(f'"{s}.{key}"' for s in sorted(owners))
        return (
            f'get("{key}") does not say which step published it, and '
            f"{len(owners)} did. Ask for one of: {suggestion}"
        )
    return (
        f'get("{key}") does not say which step published it. The form is '
        f'get("<step>.{key}") -- context is keyed by the step that wrote it, so '
        f"two steps can publish the same name without either losing it"
    )


def _not_visible(step: str, visible: list[str]) -> str:
    if not visible:
        return (
            f'no context is visible to this step, so "{step}" cannot be read. '
            f"Add it to this step's depends_on in the workflow"
        )
    return (
        f'step "{step}" is not visible to this step. Visible: '
        f"{', '.join(visible)}. Add \"{step}\" to depends_on"
    )


# --------------------------------------------------------------------------
# Writing
# --------------------------------------------------------------------------

_pending: dict[str, Any] = {}
_written = False
_warned = False


def set(*mapping: Mapping[str, Any], **values: Any) -> None:  # noqa: A001
    """Publish values for the steps that depend on this one.

        context.set(name="Daniel")
        context.set(label="Nome")
        context.set({"partitions": ["2026-09-06", "2026-09-07"]})

    Calls MERGE. Those two lines leave both ``name`` and ``label`` published;
    setting the same key twice keeps the last write. Replacing wholesale would
    let a helper that publishes one key silently erase what the caller
    published, which is the kind of bug that only shows up in the step after.

    There is no step argument, and that is what makes the isolation structural:
    the only namespace a process can write is its own.

    Nothing reaches the disk until the process exits. A step that dies hard
    publishes nothing -- which is correct, because a crashed step's output
    described work that did not finish.
    """
    if _written:
        raise ContextError(
            "context.set() was called after the context was already written. That "
            "happens when set() runs inside an atexit handler registered before "
            "this module was imported; move the call into your program's normal flow"
        )

    merged: dict[str, Any] = {}
    for m in mapping:
        if not isinstance(m, Mapping):
            raise ContextError(
                f"context.set() takes keyword arguments or a mapping, and got "
                f"{type(m).__name__}"
            )
        merged.update(m)
    merged.update(values)

    for key, value in merged.items():
        if not isinstance(key, str):
            raise ContextError(
                f"context key {key!r} is a {type(key).__name__}, and JSON object keys "
                f"are strings. It would come back as \"{key}\" and the lookup would miss"
            )
        _check_serializable(key, value)

    if not _enabled():
        return

    _pending.update(merged)
    _check_size()


def published() -> Mapping[str, Any]:
    """What this step has published so far, for a test or an assertion."""
    return dict(_pending)


def _check_serializable(key: str, value: Any) -> None:
    try:
        json.dumps(value)
    except (TypeError, ValueError) as exc:
        hint = ""
        if hasattr(value, "isoformat"):
            hint = ". For a date or a datetime, pass .isoformat()"
        raise ContextError(
            f"context key {key!r} holds a {type(value).__name__}, which is not JSON: "
            f"{exc}{hint}"
        ) from exc


def _check_size() -> None:
    encoded = json.dumps(_pending, separators=(",", ":"), sort_keys=True)
    size = len(encoded.encode("utf-8"))
    if size <= MAX_BYTES:
        return

    # Naming the largest keys is the useful half: "4.6 KB is too big" leaves the
    # caller to find which value it was.
    largest = sorted(
        ((k, len(json.dumps(v, separators=(",", ":")).encode("utf-8"))) for k, v in _pending.items()),
        key=lambda kv: -kv[1],
    )[:3]
    culprits = ", ".join(f"{k} ({n} bytes)" for k, n in largest)
    raise TooLarge(
        f"the context is {size} bytes and the ceiling is {MAX_BYTES}. It travels in the "
        f"container's termination message, which the platform TRUNCATES rather than "
        f"refuses -- so an oversized object arrives cut in half and reads downstream as "
        f"'this step published nothing'. Largest: {culprits}. Publish a reference (a "
        f"path, a watermark, a count) and leave the data where it is"
    )


# --------------------------------------------------------------------------
# Running outside Brevis
# --------------------------------------------------------------------------


def _enabled() -> bool:
    """Whether anything published will actually be read.

    Said once, at INFO, and not raised. A script that cannot be run by hand
    cannot be developed -- and a set() that quietly did nothing in production is
    exactly the silent failure this library is against, so the line is what
    tells the two apart.
    """
    global _warned
    if os.environ.get(ENV_OUTPUT):
        return True
    # No output path but a run id means an executor with no return path of its
    # own -- a host the engine did not create. What is published then goes out
    # as a marked line on stdout, which every executor that streams logs has.
    if os.environ.get(ENV_RUN_ID):
        return True
    if not _warned:
        _warned = True
        _log.info(
            "brevis: not running under Brevis (neither %s nor %s is set), so "
            "context.set() is a no-op. Under the engine it publishes to the steps "
            "that depend on this one",
            ENV_OUTPUT,
            ENV_RUN_ID,
        )
    return False


def _flush() -> None:
    """Write once, at exit. Registered on import, so a caller need not."""
    global _written
    if _written or not _pending:
        return
    path = os.environ.get(ENV_OUTPUT)
    if not path:
        # No file: publish on stdout instead. There is deliberately no second
        # check for the run id here -- ``_enabled()`` is the one gate, and
        # nothing reaches ``_pending`` outside the engine, so a guard here would
        # be a branch no test could ever reach. A mutation proved exactly that:
        # removing it changed nothing anywhere.
        _written = True
        _announce(json.dumps(_pending, separators=(",", ":"), sort_keys=True))
        return

    _written = True
    payload = json.dumps(_pending, separators=(",", ":"), sort_keys=True)

    try:
        # /dev/termination-log is a character device: it is written directly,
        # never renamed into place. A temp-file-and-rename would fail there, and
        # atomicity is not needed because the engine reads this exactly once,
        # after the process is gone.
        with open(path, "w", encoding="utf-8") as fh:
            fh.write(payload)
    except OSError as exc:
        # Loud, and on stderr, because the step is about to be reported as
        # succeeded while the step after it will not find what it needs.
        _log.error(
            "brevis: could not publish the context to %s: %s. The steps depending on "
            "this one will not see it",
            path,
            exc,
        )


def _announce(payload: str) -> None:
    """Publish on stdout, for an executor with no return path of its own.

    ONE write, flushed.

    One write rather than ``print()``, because ``print()`` does two -- the text,
    then the newline -- and stdout is shared with everything else the step logs.
    A second thread writing in between splits the marker across two lines, and
    the engine's scanner breaks on lines, so what it sees is not a marker at all.

    Flushed because a step killed by a timeout or a cancel after publishing
    would otherwise lose the line entirely.

    The file is preferred wherever there is one. Two roads carrying the same
    thing is what drifts, and a file is the road the platform vouches for.
    """
    sys.stdout.write(f'{MARKER}{{"type":"context","value":{payload}}}\n')
    sys.stdout.flush()


atexit.register(_flush)


def _reset_for_tests(output: str | None = None) -> str:
    """Clear the module's state between tests, and point the output somewhere.

    Exported with a leading underscore because it is not part of the contract:
    a real step is one process, one step, one write, and has no reason to reset.
    """
    global _written, _warned
    _pending.clear()
    _written = False
    _warned = False
    if output is None:
        output = os.path.join(tempfile.mkdtemp(), "output")
    os.environ[ENV_OUTPUT] = output
    return output

File: sdk/context/context.go
// Package context is what one step tells the next.
//
//	import bctx "github.com/AreteAcademy/brevis/sdk/context"
//
//	bucket, err := bctx.String("extract.bucket")
//	err = bctx.Set("rows", 48213)
//
// It is the Go half of a contract that is not Go's: the same two environment
// variables the Python library reads, in the same wire format. A Go step and a
// Python step in one workflow exchange context without either knowing which
// wrote it, and a `bash` step with `jq` joins in without a library at all.
//
// # It depends on nothing
//
// Not on the rest of this SDK, not on a driver, not on the network. Importing
// it costs `encoding/json`, `os` and `strings`. A fetcher that wants context
// and nothing else pays for context and nothing else, which is the same rule
// every driver in this module follows.
//
// # How it reaches the engine
//
// It does not. There is no API call, no token and no port: the engine hands a
// step BREVIS_INPUT when it starts it and reads BREVIS_OUTPUT back when it
// ends. Under Kubernetes that path is /dev/termination-log, which the engine
// already reads to get the exit code.
package context

import (
	"encoding/json"
	"fmt"
	"os"
	"sort"
	"strings"
	"sync"
)

// The transport. A consumer never names these; they are here because a bash
// step does.
const (
	EnvInput  = "BREVIS_INPUT"
	EnvOutput = "BREVIS_OUTPUT"
)

// MaxBytes is how much a step may publish, and it is the platform's number
// rather than this SDK's: the container's termination message is truncated at
// 4096 bytes.
//
// Inheriting a ceiling is stronger than enforcing a policy. Four kilobytes hold
// a watermark, a count or a list of partitions; they do not hold the rows, and
// that is the boundary this feature is for.
const MaxBytes = 4096

// ---------------------------------------------------------------------------
// Reading
// ---------------------------------------------------------------------------

var (
	once     sync.Once
	incoming map[string]map[string]any
	inputErr error
)

func read() (map[string]map[string]any, error) {
	once.Do(func() {
		raw := strings.TrimSpace(os.Getenv(EnvInput))
		if raw == "" {
			incoming = map[string]map[string]any{}
			return
		}
		if err := json.Unmarshal([]byte(raw), &incoming); err != nil {
			inputErr = fmt.Errorf("%s is not the expected JSON. The engine writes it, so "+
				"this is a version mismatch rather than anything you did: %w", EnvInput, err)
		}
	})
	return incoming, inputErr
}

// Value reads one value published by a step this one depends on.
//
// The key is ALWAYS qualified by the step that wrote it -- "extract.bucket" --
// and a bare "bucket" is refused naming the steps that published it.
//
// That is not ceremony. Steps are isolated: `extract` and `transform` each
// write only their own namespace, so both can publish `bucket` without either
// losing it. A bare key throws that away the moment two of them do, and
// searching and picking one is precedence by accident -- it works until it
// silently does not, and somebody spends an afternoon on a value that came from
// the wrong step.
//
// A key that is absent returns ok=false with no error. A STEP that is absent is
// an error, because it means the pipeline does not say what the code assumes.
func Value(key string) (any, bool, error) {
	step, field, ok := strings.Cut(key, ".")
	if !ok || step == "" || field == "" {
		return nil, false, unqualified(key)
	}

	all, err := read()
	if err != nil {
		return nil, false, err
	}
	values, present := all[step]
	if !present {
		return nil, false, notVisible(step, all)
	}
	v, found := values[field]
	return v, found, nil
}

// String reads a text value. A value that is not a string is an error naming
// the type, because a silent "" downstream is a flag with no argument.
func String(key string) (string, error) {
	v, ok, err := Value(key)
	if err != nil || !ok {
		return "", err
	}
	s, isString := v.(string)
	if !isString {
		return "", fmt.Errorf("context %q holds a %T, not a string", key, v)
	}
	return s, nil
}

// Int reads a whole number.
//
// JSON has one number type, so an integer arrives as a float64. A value with a
// fractional part is an error rather than a truncation: silently turning 1.5
// into 1 is the class of difference this project has already paid for once, in
// ingestion_id.
func Int(key string) (int, error) {
	v, ok, err := Value(key)
	if err != nil || !ok {
		return 0, err
	}
	f, isNumber := v.(float64)
	if !isNumber {
		return 0, fmt.Errorf("context %q holds a %T, not a number", key, v)
	}
	if f != float64(int(f)) {
		return 0, fmt.Errorf("context %q holds %v, which is not a whole number", key, f)
	}
	return int(f), nil
}

// Into decodes one step's whole published object into a struct or a map.
func Into(step string, dst any) error {
	all, err := read()
	if err != nil {
		return err
	}
	values, present := all[step]
	if !present {
		return notVisible(step, all)
	}
	raw, err := json.Marshal(values)
	if err != nil {
		return err
	}
	return json.Unmarshal(raw, dst)
}

func unqualified(key string) error {
	all, _ := read()
	var owners []string
	for step, values := range all {
		if _, ok := values[key]; ok {
			owners = append(owners, step)
		}
	}
	sort.Strings(owners)
	if len(owners) > 0 {
		quoted := make([]string, len(owners))
		for i, s := range owners {
			quoted[i] = fmt.Sprintf("%q", s+"."+key)
		}
		return fmt.Errorf("%q does not say which step published it, and %d did. Ask for "+
			"one of: %s", key, len(owners), strings.Join(quoted, ", "))
	}
	return fmt.Errorf("%q does not say which step published it. The form is \"<step>.%s\" "+
		"-- context is keyed by the step that wrote it, so two steps can publish the same "+
		"name without either losing it", key, key)
}

func notVisible(step string, all map[string]map[string]any) error {
	visible := make([]string, 0, len(all))
	for id := range all {
		visible = append(visible, id)
	}
	sort.Strings(visible)
	if len(visible) == 0 {
		return fmt.Errorf("no context is visible to this step, so %q cannot be read. Add it "+
			"to this step's depends_on in the workflow", step)
	}
	return fmt.Errorf("step %q is not visible to this step. Visible: %s. Add %q to "+
		"depends_on", step, strings.Join(visible, ", "), step)
}

// ---------------------------------------------------------------------------
// Writing
// ---------------------------------------------------------------------------

var (
	mu        sync.Mutex
	published = map[string]any{}
)

// Set publishes one value for the steps that depend on this one.
//
//	bctx.Set("rows", 48213)
//	bctx.Set("watermark", "2026-09-07T03:00:00Z")
//
// Calls MERGE: those two leave both keys published, and setting the same key
// twice keeps the last write. Replacing wholesale would let a helper that
// publishes one key silently erase what its caller published -- a bug that only
// appears in the step after this one.
//
// There is no step argument, and that is what makes the isolation structural:
// the only namespace a process can write is its own.
//
// # Why this writes immediately, and Python's does not
//
// Python defers to `atexit`, so one call reaches the disk at the end. Go has no
// atexit, and the alternative -- requiring a `defer Flush()` -- makes a
// forgotten line into silent data loss, which is the failure this whole feature
// exists to prevent. So each Set rewrites the file, which at four kilobytes
// costs nothing.
//
// The visible difference is that a Go step killed mid-run publishes what it had
// set so far, and a Python one publishes nothing. Neither is read: when a step
// fails, nothing below it runs, and its retry republishes from scratch.
func Set(key string, value any) error {
	if key == "" {
		return fmt.Errorf("context key is empty")
	}
	if err := serializable(key, value); err != nil {
		return err
	}

	mu.Lock()
	defer mu.Unlock()

	previous, had := published[key]
	published[key] = value

	encoded, err := encode()
	if err != nil {
		return err
	}
	if len(encoded) > MaxBytes {
		// The message is built BEFORE the rollback: it names what is holding the
		// bytes, and rolling back first empties the very map it reads. A test
		// caught that -- the refusal arrived saying "Holding: " with nothing
		// after it, which is the half of the message that makes it useful.
		err := tooLarge(len(encoded))

		// Put it back, so a refused Set leaves nothing behind: otherwise the
		// next Set fails on a key the caller believes it never set.
		if had {
			published[key] = previous
		} else {
			delete(published, key)
		}
		return err
	}
	return write(encoded)
}

// SetAll publishes several values at once, and is the same merge.
func SetAll(values map[string]any) error {
	keys := make([]string, 0, len(values))
	for k := range values {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	for _, k := range keys {
		if err := Set(k, values[k]); err != nil {
			return err
		}
	}
	return nil
}

// Published is what this step has published so far, for a test or a log line.
func Published() map[string]any {
	mu.Lock()
	defer mu.Unlock()
	out := make(map[string]any, len(published))
	for k, v := range published {
		out[k] = v
	}
	return out
}

func encode() ([]byte, error) {
	b, err := json.Marshal(published)
	if err != nil {
		return nil, fmt.Errorf("the context is not serializable: %w", err)
	}
	return b, nil
}

func serializable(key string, value any) error {
	if _, err := json.Marshal(value); err != nil {
		return fmt.Errorf("context key %q holds a %T, which is not JSON: %w", key, value, err)
	}
	return nil
}

func tooLarge(size int) error {
	mu2 := make([]string, 0, len(published))
	for k, v := range published {
		b, _ := json.Marshal(v)
		mu2 = append(mu2, fmt.Sprintf("%s (%d bytes)", k, len(b)))
	}
	sort.Strings(mu2)
	return fmt.Errorf("the context is %d bytes and the ceiling is %d. It travels in the "+
		"container's termination message, which the platform TRUNCATES rather than refuses "+
		"-- so an oversized object arrives cut in half and reads downstream as 'this step "+
		"published nothing'. Holding: %s. Publish a reference (a path, a watermark, a "+
		"count) and leave the data where it is",
		size, MaxBytes, strings.Join(mu2, ", "))
}

// write puts the whole object where the engine will read it.
//
// Nothing happens outside Brevis: no output path means nothing is going to read
// it, and a step has to stay runnable by hand. Unlike the Python library this
// does not log a line, because a Go fetcher under sdk.Run already reports its
// own context and a second voice there would be noise.
func write(encoded []byte) error {
	path := os.Getenv(EnvOutput)
	if path == "" {
		return nil
	}
	// Written directly, never renamed into place: /dev/termination-log is
	// provided by the platform, and a rename over it would fail. Atomicity is
	// not needed because the engine reads it once, after this process is gone.
	if err := os.WriteFile(path, encoded, 0o600); err != nil {
		return fmt.Errorf("publishing the context to %s: %w. The steps depending on this "+
			"one will not see it", path, err)
	}
	return nil
}

// resetOnce exists for this package's own tests: a real step is one process
// reading one input, and has no reason to read it twice.
func resetOnce() sync.Once { return sync.Once{} }
